service/src/main/java/bio/terra/workspace/service/iam/SamService.java

```diff
@@ -884,15 +889,38 @@ private void addPrivateResourcePolicies(
    * Fetch the email of a user's pet service account in a given project. This request to Sam will
    * create the pet SA if it doesn't already exist.
    */
-  public String getOrCreatePetSaEmail(String projectId, AuthenticatedUserRequest userRequest) {
+  public String getOrCreatePetSaEmail(String projectId, AuthenticatedUserRequest userRequest)
+      throws InterruptedException {
     GoogleApi googleApi = samGoogleApi(userRequest.getRequiredToken());
     try {
-      return googleApi.getPetServiceAccount(projectId);
+      return SamRetry.retry(() -> googleApi.getPetServiceAccount(projectId));
     } catch (ApiException apiException) {
       throw SamExceptionFactory.create("Error getting pet service account from Sam", apiException);
     }
   }
 
+  // TODO(PF-991): This is a temporary workaround to support disabling pet service account
+  //  self-impersonation without having user credentials available. When we stop granting this
+  //  permission directly to users and their pets, this method should be deleted.
+  /**
+   * Construct the email of an arbitrary user's pet service account in a given project. Unlike
+   * {@code getOrCreatePetSaEmail}, this will not create the underlying service account. It may
+   * return the email of a service account which does not exist.
+   */
+  public ServiceAccountName constructUserPetSaEmail(
+      String projectId, String userEmail, AuthenticatedUserRequest userRequest)
+      throws InterruptedException {
+    UsersApi usersApi = samUsersApi(userRequest.getRequiredToken());
+    try {
+      String subjectId = usersApi.getUserIds(userEmail).getUserSubjectId();
+      String saEmail = String.format("pet-%s@%s.iam.gserviceaccount.com", subjectId, projectId);
+      return SamRetry.retry(
```

Comment:
nit: Do you need `SamRetry` here? I thought this wasn't actually calling SAM, just building a string.